dwi/dif_eddy_trac_bids.py

```diff
@@ -0,0 +1,555 @@
+    preproc.inputs.order = pe_key    
+
+    def rotate_bvecs(bvec_file, par_file):
+        """ Rotates bvecs """
+        import os
+        import numpy as np
+        pars = np.genfromtxt(par_file)
+        bvecs = np.genfromtxt(bvec_file)
+        new_bvecs = []
+        rotfunc = lambda x: np.array([[np.cos(x), np.sin(x)],
+                                      [-np.sin(x), np.cos(x)]])
+        for idx, vector in enumerate(bvecs):
+            par = pars[idx]
+            Rx = np.eye(3)
+            Rx[1:3, 1:3] = rotfunc(par[3])
+            Ry = np.eye(3)
+            Ry[(0, 0, 2, 2), (0, 2, 0, 2)] = rotfunc(par[4]).ravel()
+            Rz = np.eye(3)
+            Rz[0:2, 0:2] = rotfunc(par[5])
+            R = np.linalg.inv(Rx.dot(Ry.dot(Rz)))
+            new_bvecs.append(R.dot(vector.T).tolist())
+        new_bvec_file = os.path.join(os.getcwd(), 'rotated.bvecs')
+        np.savetxt(new_bvec_file, new_bvecs, '%.10f %.10f %.10f')
+        return new_bvec_file
+
+    rotate = Node(Function(input_names=['bvec_file', 'par_file'],
+                            output_names=['bvec_file'],
+                            function=rotate_bvecs), name='rotate')
+
+    wf.connect(preproc, 'merged_bvecs', rotate, 'bvec_file')
+    wf.connect(eddy, 'out_parameter', rotate, 'par_file')
+    
+    """
+    TRACULA
+    """
+
+    def run_prep(sid, template, nifti, bvec, bval, tracula_dir):
+        """ Runs trac-all from command line, could convert into interface """
+        from glob import glob
+        import oss
+        from string import Template
+        with open(template, 'rt') as fp:
+            tpl = Template(fp.read())
+        out = tpl.substitute(subjects=sid, bvec=bvec, bval=bval, niftis=nifti)
+        config_file = os.path.join(os.getcwd(), 'config_%s' % sid)
+        with open(config_file, 'wt') as fp:
+            fp.write(out)
+        from nipype.interfaces.base import CommandLine
+        from nipype.pipeline.engine import Node
+        node = Node(CommandLine('trac-all -prep -c %s -no-isrunning -noqa' % config_file, 
+                                terminal_output='allatonce'),
+                    name='trac-prep-%s' % sid)
+        node.base_dir = os.getcwd()
+        node.run()
+        dwi_file = os.path.join(tracula_dir, sid, 'dmri', 'dwi.nii.gz')
+        return sid, config_file, dwi_file
+
+    node1 = Node(Function(input_names=['sid', 'template', 'nifti', 'bvec', 'bval', 'tracula_dir'],
+                          output_names=['sid', 'config_file', 'dwi_file'], 
+                          function=run_prep),
+                 name='trac-prep')
+    
+    def run_bedpost(sid, tracula_dir, dwi_file):
+        """ """
+        import os
+        import shutil
+        from nipype.interfaces.base import CommandLine
+        pwd = os.getcwd()
+        os.chdir(os.path.join(tracula_dir, sid))
+        bedpost = CommandLine('bedpostx_gpu dmri -NJOBS 1 --model=2 --rician', terminal_output='allatonce')
```

Comment:
note that these parameters may need to be modified depending on whether we are processing prisma or trio data.